### PR TITLE
Added ConfigValidation interface

### PR DESCRIPTION
**Pull request by mohitsoni**

### CODE REVIEW

File: src/main/java/org/apache/mesos/config/ConfigurationValidation.java [hunk 1]
```diff
@@ -0,0 +1,19 @@
+package org.apache.mesos.config;
+
+import java.util.List;
+
+/**
+ * The {@code ConfigurationValidation} interface should be implemented by any class
+ * which intend to validate a new {@code Configuration} w.r.t. an old {@code Configuration}.
+ */
+public interface ConfigurationValidation {
```
Comment (gabrielhartmann) [inline src/main/java/org/apache/mesos/config/ConfigurationValidation.java]: ConfigurationValidator

File: src/main/java/org/apache/mesos/config/ConfigurationValidationError.java [hunk 2]
```diff
@@ -0,0 +1,27 @@
+package org.apache.mesos.config;
+
+/**
+ * Signals a certain configurationField is incorrectly configured.
+ */
+public class ConfigurationValidationError {
+    private final String configurationField;
+    private final String message;
+
+    public ConfigurationValidationError(String configField, String message) {
```
Comment (gabrielhartmann) [inline src/main/java/org/apache/mesos/config/ConfigurationValidationError.java]: Should also contain `String configValue`?

Comment (mohitsoni) [inline src/main/java/org/apache/mesos/config/ConfigurationValidationError.java]: Good point!

File: src/main/java/org/apache/mesos/config/ConfigurationValidation.java (hunk 1 above)
Comment (nickbp) [inline src/main/java/org/apache/mesos/config/ConfigurationValidation.java]: Nit: ConfigurationValidator?

File: src/main/java/org/apache/mesos/config/ConfigurationValidationException.java [hunk 3]
```diff
@@ -0,0 +1,20 @@
+package org.apache.mesos.config;
```
Comment (nickbp) [inline src/main/java/org/apache/mesos/config/ConfigurationValidationException.java]: Looks like this exception isn't being used. Remove?

File: src/main/java/org/apache/mesos/config/ConfigurationValidation.java [hunk 4]
```diff
@@ -0,0 +1,19 @@
...
+    /**
+     * Returns {@code List} of {@code ConfigurationValidationError} for the newly supplied
+     * {@code Configuration} object.
+     *
+     * @param oldConfig Currently persisted Configuration.
+     * @param newConfig Proposed new Configuration
+     * @return List of errors
```
Comment (nickbp) [inline src/main/java/org/apache/mesos/config/ConfigurationValidation.java]: "List of errors, or an empty list if validation passed"

**Comment by nickbp** [comment]
For 'validate the new config, regardless of the old config' (eg 'ram should be non-zero'), would that just be implemented as a ConfigValidator that ignores the oldConfig parameter?

File: src/main/java/org/apache/mesos/config/DefaultConfigurationValidator.java [hunk 5]
```diff
@@ -0,0 +1,27 @@
...
+import java.util.List;
+
+/**
+ * Default Configuration validator, which validates a given new {@code Configuration} w.r.t.
+ * an old {@code Configuration}.
+ */
+public class DefaultConfigurationValidator {
+    private final Collection<ConfigurationValidation> validations = new ArrayList<>();
+
+    public DefaultConfigurationValidator(final Collection<ConfigurationValidation> validations) {
```
Comment (nickbp) [inline src/main/java/org/apache/mesos/config/DefaultConfigurationValidator.java]: Would be useful to have an alternate constructor of the form:

```
Constructor(ConfigurationValidator... validators) {
  this(Arrays.asList(validators));
}
```

Then callers could create a new validator like this:

```
new DefaultConfigurationValidator(
  new MyValidatorA(),
  new MyValidatorB(),
  ...);
```

File: src/main/java/org/apache/mesos/config/ConfigurationValidationException.java (hunk 3 above)
Comment (mohitsoni) [inline src/main/java/org/apache/mesos/config/ConfigurationValidationException.java]: This is intended for SDK users.

File: src/main/java/org/apache/mesos/config/DefaultConfigurationValidator.java (hunk 5 above)
Comment (mohitsoni) [inline src/main/java/org/apache/mesos/config/DefaultConfigurationValidator.java]: Good suggestion.

File: src/main/java/org/apache/mesos/config/ConfigurationValidationException.java (hunk 3 above)
Comment (gabrielhartmann) [inline src/main/java/org/apache/mesos/config/ConfigurationValidationException.java]: I agree with @nickbp.  If it's not being used yet, we should remove it.

**Comment by mohitsoni** [comment]
@nickbp Regarding new config only validation. See here: https://github.com/mesosphere/dcos-cassandra-service/blob/master/cassandra-scheduler/src/main/java/com/mesosphere/dcos/cassandra/scheduler/config/ConfigValidator.java#L32

**Comment by gabrielhartmann** [comment]
@mohitsoni:  Not sure I understand what you're referring to in that link.  I could be wrong but I think @nickbp is talking about the case where there is no `oldConfig`.  Some validation should occur even on first startup to guard against things Universe validation can't catch.

**Comment by mohitsoni** [comment]
@gabrielhartmann That was what I was trying to point out. That link demonstrates that we verify  `seeds < servers` always. Presence of an old config doesn't matter, in that example.

**Comment by mohitsoni** [comment]
We can design(overdesign?) it, by having two separate validate methods to make the intention more explicit:
`validate(newConfig)`
`validate(oldConfig, newConfig)`

**Comment by gabrielhartmann** [comment]
@mohitsoni:  I see, so on this line we're comparing `newConfig` to `newConfig`: https://github.com/mesosphere/dcos-cassandra-service/blob/master/cassandra-scheduler/src/main/java/com/mesosphere/dcos/cassandra/scheduler/config/ConfigValidator.java#L35

That's a little confusing from an interface perspective.  It would be nice to be able to specify Validations that take a single Configuration to make clear that it isn't doing a comparison to another configuration.

**Comment by mohitsoni** [comment]
@gabrielhartmann Please see my earlier comment.

**Comment by nickbp** [comment]
TBH I'm ok with using the same type/interface for both self-validation and cross-validation, was just thinking it'd be good to document the two possibilities.

**Comment by gabrielhartmann** [comment]
Ok, just a thought.  Can be changed alter if needed.  Don't block on my comments.  LGTM.

**Comment by mohitsoni** [comment]
@nickbp That was my un-documented intent. But, as I pointed out above, we can make things more explicit by having dedicated methods. It's a design choice.

**Comment by nickbp** [comment]
LGTM

**Pull request merged**